File: dsa-problems/Bitwise/solutions/BIT-008-maximize-or-k-picks.py
import sys

def maximize_or_with_k_picks(a: list[int], k: int) -> int:
    n = len(a)
    # Optimization: If K is large enough, we can set all possible bits
    if k >= 30:
        total = 0
        for x in a:
            total |= x
        return total
        
    current_or = 0
    used = [False] * n
    
    for _ in range(k):
        best_or = -1
        best_idx = -1
        
        for i in range(n):
            if not used[i]:
                new_or = current_or | a[i]
                if new_or > best_or:
                    best_or = new_or
                    best_idx = i
                    
        # If we found something (which we always should if k <= n)
        if best_idx != -1:
            current_or = best_or
            used[best_idx] = True
            
    return current_or

def main():
    input_data = sys.stdin.read()
    data = input_data.split()
    if not data: return
    
    ptr = 0
    n = int(data[ptr]); ptr += 1
    a = []
    for _ in range(n):
        a.append(int(data[ptr])); ptr += 1
    
    k = int(data[ptr]); ptr += 1
    
    result = maximize_or_with_k_picks(a, k)
    print(result)

if __name__ == "__main__":
    main()

File: dsa-problems/Bitwise/solutions/java/BIT-008-maximize-or-k-picks.java
import java.util.*;

class Solution {
    public long maximizeOrWithKPicks(int[] a, int k) {
        int n = a.length;
        // Optimization: 30 bits max. If k >= 30, we can collect all bits.
        if (k >= 30) {
            long totalOr = 0;
            for (int x : a) totalOr |= x;
            return totalOr;
        }

        long currentOr = 0;
        boolean[] used = new boolean[n];

        for (int step = 0; step < k; step++) {
            long bestOr = -1;
            int bestIdx = -1;

            for (int i = 0; i < n; i++) {
                if (!used[i]) {
                    long newOr = currentOr | a[i];
                    if (newOr > bestOr) {
                        bestOr = newOr;
                        bestIdx = i;
                    }
                }
            }

            if (bestIdx != -1) {
                currentOr = bestOr;
                used[bestIdx] = true;
            }
        }
        return currentOr;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int[] a = new int[n];
        for (int i = 0; i < n; i++) a[i] = sc.nextInt();
        int k = sc.nextInt();

        Solution solution = new Solution();
        System.out.println(solution.maximizeOrWithKPicks(a, k));
        sc.close();
    }
}
